### DISCUSSION TITLE
Async UART echo received buffer issue

### DISCUSSION THREAD

**Discussion by krupis** [zephyrproject-rtos/zephyr#64121] [General]
**Is your feature request related to a problem? Please describe.**
I am learning about Zephyr RTOS and nRF52 microcontrollers. I have been following the Nordic DevAcademy basic tutorial:
https://academy.nordicsemi.com/courses/nrf-connect-sdk-fundamentals/lessons/lesson-4-serial-communication-uart/topic/uart-driver/

The tutorial includes lesson about serial communication (UART).

One of the most important things for me when I am working with UART is to be able to print out whole received buffer. Then I can look at the data and do whatever I want with it. However, this simple task is somehow proving to be difficult using ASYNC UART method.

**Describe the solution you'd like**
A simple way to print out whole received buffer

**Additional context**
My main.c:
```
#include <zephyr/kernel.h>
#include <zephyr/drivers/gpio.h>
#include "stdio.h"
#include "my_gpio.h"
#include "zephyr/drivers/uart.h"

const struct device *uart = DEVICE_DT_GET(DT_NODELABEL(uart0));

static uint8_t tx_buf[] = {"nRF Connect SDK Fundamentals Course \n\r"};
static uint8_t rx_buf[10] = {0}; // A buffer to store incoming UART data

const struct uart_config uart_cfg = {
	.baudrate = 115200,
	.parity = UART_CFG_PARITY_NONE,
	.stop_bits = UART_CFG_STOP_BITS_1,
	.data_bits = UART_CFG_DATA_BITS_8,
	.flow_ctrl = UART_CFG_FLOW_CTRL_NONE};

static void uart_cb(const struct device *dev, struct uart_event *evt, void *user_data)
{
	switch (evt->type)
	{

	case UART_TX_DONE:
		printf("transmission complete \n");
		break;

	case UART_TX_ABORTED:
		// do something
		break;

	case UART_RX_RDY:
                printf("Data received = %s \n",evt->data.rx_buf);
		break;

	case UART_RX_BUF_REQUEST:
		printf("requesting buffer \n");
		// do something
		break;

	case UART_RX_BUF_RELEASED:
		printf("buffer released \n");
		// do something
		break;

	case UART_RX_DISABLED:
		printf("rx disabled \n");
		uart_rx_enable(dev, rx_buf, sizeof(rx_buf), 100);
		break;

	case UART_RX_STOPPED:
		// do something
		break;

	default:
		break;
	}
}

int main(void)
{

	if (!device_is_ready(uart))
	{
		return;
	}

	int err;
	err = uart_callback_set(uart, uart_cb, NULL);
	if (err)
	{
		return err;
	}

	uart_rx_enable(uart, rx_buf, sizeof(rx_buf), 100);

	while (1)
	{
		k_msleep(1000);
	}
	return 0;
}

```

As you can see, from above, I tried to print out the buffer:
```
	case UART_RX_RDY:
                printf("Data received = %s \n",evt->data.rx_buf);
		break;
```
but it does not work as expect. Please see the following terminal logs: (The blue is what I send, The green is what is printed)
![image](https://github.com/zephyrproject-rtos/zephyr/assets/22237034/7264ac0f-1cc4-4dae-94c9-62bdba6245cf)

Could someone help me understand how to simply print out the data that I have sent?

**Comment by bjarki-andreasen** [comment]
Hi, You should the [logging subsystem](https://docs.zephyrproject.org/latest/services/logging/index.html#api-reference)
```
LOG_HEXDUMP_DBG(&evt->data.rx_buf.buf[evt->data.rx_buf.offset], evt->data.rx_buf.len, "Data received = ");
```
The issues you are seeing are probably a combination of not reading at the correct offset within the rx buffer, and using %s to format the data in the buffer which is not safe given the potential of no string terminator present in the received data :)

**Comment by krupis** [comment]
@bjarki-trackunit 
Hello. Thank you very much for your response. I think I am getting closer to what I want to achieve, but I still have some problems with your suggested method.

I have modified my ```UART_RX_RDY``` as you have suggested:

```
	case UART_RX_RDY:
		LOG_HEXDUMP_DBG(&evt->data.rx_buf.buf[evt->data.rx.offset], evt->data.rx.len  , "Data received: ");
		break;
```

Just notice that I have replaced ```evt->data.rx_buf.offset``` with ```evt->data.rx.offset``` as rx_buff does not have offset struct member. Same goes for ```evt->data.rx_buf.len``` , I have replaced it with ```evt->data.rx.len``` .

In my program, the rx_buf length is set to 20

See the logs below:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/22237034/d322f834-5a81-46ad-9c4f-2dafcc657f42)


As you can see from logs above, I have sucesfully received messages ```ping```, ```test```, ```hello123``` , but when I send ```long_message```. Something bad happen and I could not receive the full buffer. I think that is because ```buffer release``` was triggered because it filled up and I could not receive the full message. 

Can you please help me understand how to handle this case? The issue is related to when the buffer fills up, but I would still like to be able to receive a full message. Do you understand what I mean?
